src/librustc/traits/fulfill.rs

```diff
@@ -99,13 +104,98 @@ pub struct PendingPredicateObligation<'tcx> {
     pub stalled_on: Vec<Ty<'tcx>>,
 }
 
+/// An obligation which cannot be fulfilled in the context
+/// it was registered in, such as auto trait obligations on
+/// `impl Trait`, which require the concrete type to be
+/// available, only guaranteed after finishing type-checking.
+#[derive(Clone, Debug)]
+pub struct DeferredObligation<'tcx> {
+    pub predicate: ty::PolyTraitPredicate<'tcx>,
+    pub cause: ObligationCause<'tcx>
+}
```

Comment:
Ah, right, I got it mixed up. Still, we don't pick the default, we just don't normalize, so doing the same for `impl Trait` would be just conservative, not wrong.